[MLC-199] added 'ModelInferenceTimeCheck';

yromanyshyn opened the pull request.

noamzbr commented on deepchecks/checks/performance/model_inference_time.py:
```diff
@@ -0,0 +1,84 @@
+        Dataset.validate_dataset(dataset, check_name)
+        model_type_validation(model)
+
+        if dataset.features_columns() is None:
```
We're missing dataset.validate_model(model), and it will also take care of checking that there is inference to be done

yromanyshyn replied:
updated

noamzbr commented on deepchecks/checks/performance/model_inference_time.py:
```diff
@@ -0,0 +1,84 @@
+class ModelInferenceTimeCheck(SingleDatasetBaseCheck):
+    """Measure model average inference time (in seconds) per sample."""
+
+    NUMBER_OF_SAMPLES: t.ClassVar[int] = 1000
```
Shouldn't this appear as a check parameter in the __init__?

yromanyshyn replied:
updated

noamzbr commented on deepchecks/checks/performance/model_inference_time.py:
```diff
@@ -0,0 +1,84 @@
+            raise DeepchecksValueError(f'Check {check_name} requires dataset with the features!')
+
+        prediction_method = model.predict # type: ignore
+        df = dataset.features_columns()[:self.NUMBER_OF_SAMPLES] # type: ignore
```
we still want to be able to function when there are less then NUMBER_OF_SAMPLES in data

yromanyshyn replied:
but it still will be working if `len(df) < NUMBER_OF_SAMPLES`

example:
```
import pandas as pd
df = pd.DataFrame.from_dict({"a": list(range(100))})
df[:10000] # works, will return all data
```

yromanyshyn replied:
oh, I understand (stupid mistake) - fixed it

noamzbr commented on deepchecks/checks/performance/model_inference_time.py:
```diff
@@ -0,0 +1,84 @@
+        result = result / self.NUMBER_OF_SAMPLES
+
+        return CheckResult(value=result, check=type(self), display=(
+            'Average model inference time for one sample (in seconds) '
```
```suggestion
            'Average model inference time for one sample (in seconds): {format_number(result, floating_point=8)} '
```

noamzbr commented on deepchecks/checks/performance/model_inference_time.py:
```diff
@@ -0,0 +1,88 @@
+        prediction_method = model.predict # type: ignore
+        df = test_dataset.features_columns()
+
+        assert df is not None, "Internal Error! 'dataset._features' var was not initialized!"  # pylint: disable=inconsistent-quotes
```
really, what they failed on?

yromanyshyn replied:
`validate_dataset` raised error

test logs:
```
FAILED tests/checks/integrity/new_label_test.py::test_no_new_label - deepchecks.utils.DeepchecksValueError: Check NewLabelTra...
FAILED tests/checks/integrity/new_label_test.py::test_new_label - deepchecks.utils.DeepchecksValueError: Check NewLabelTrainT...
FAILED tests/checks/integrity/new_label_test.py::test_missing_label - deepchecks.utils.DeepchecksValueError: Check NewLabelTr...
FAILED tests/checks/integrity/new_label_test.py::test_missing_new_label - deepchecks.utils.DeepchecksValueError: Check NewLab...
FAILED tests/checks/methodology/date_leakage_test.py::test_dates_from_val_in_train - deepchecks.utils.DeepchecksValueError: C...
FAILED tests/checks/methodology/date_leakage_test.py::test_limit_dates_from_val_in_train - deepchecks.utils.DeepchecksValueEr...
FAILED tests/checks/methodology/date_leakage_test.py::test_no_dates_from_val_in_train - deepchecks.utils.DeepchecksValueError...
FAILED tests/checks/methodology/date_leakage_test.py::test_dates_from_val_before_train - deepchecks.utils.DeepchecksValueErro...
FAILED tests/checks/methodology/date_leakage_test.py::test_no_dates_from_val_before_train - deepchecks.utils.DeepchecksValueE...
FAILED tests/checks/methodology/date_leakage_test.py::test_nan - deepchecks.utils.DeepchecksValueError: Check DateTrainTestLe...
FAILED tests/checks/methodology/date_leakage_test.py::test_condition_fail_on_overlap - deepchecks.utils.DeepchecksValueError:...
FAILED tests/checks/methodology/date_leakage_test.py::test_condition_on_overlap - deepchecks.utils.DeepchecksValueError: Chec...
FAILED tests/checks/methodology/date_leakage_test.py::test_condition_fail_on_duplicates - deepchecks.utils.DeepchecksValueErr...
FAILED tests/checks/methodology/date_leakage_test.py::test_condition_pass_on_duplicates - deepchecks.utils.DeepchecksValueErr...
FAILED tests/checks/methodology/index_train_val_leakage_test.py::test_indexes_from_val_in_train - deepchecks.utils.Deepchecks...
FAILED tests/checks/methodology/index_train_val_leakage_test.py::test_limit_indexes_from_val_in_train - deepchecks.utils.Deep...
FAILED tests/checks/methodology/index_train_val_leakage_test.py::test_no_indexes_from_val_in_train - deepchecks.utils.Deepche...
FAILED tests/checks/methodology/index_train_val_leakage_test.py::test_nan - deepchecks.utils.DeepchecksValueError: Check Inde...
FAILED tests/checks/methodology/index_train_val_leakage_test.py::test_condition_leakage_fail - deepchecks.utils.DeepchecksVal...
FAILED tests/checks/methodology/index_train_val_leakage_test.py::test_condition_leakage_passesl - deepchecks.utils.Deepchecks...
```

noamzbr replied:
I'm still missing something - if they are failing when you start checking for None DataFrames in validate_dataset, it means these tests are passing None DataFrames. Is that the case?

yromanyshyn replied:
when I added the next if statement to the `validate_dataset` method they started failing:

```
if obj.features_columns() is None:
    raise DeepchecksValueError(f'Check {check_name} requires dataset with the features!')
```

noamzbr replied:
Oh now I understand. Yes it makes sense - not all checks require features. I makes sense to have a separate validation for the presence of features.

matanper replied:
+1 for Yurii

noamzbr commented on deepchecks/checks/performance/model_inference_time.py:
```diff
@@ -0,0 +1,92 @@
+
+        result = timeit.timeit(
+            'predict(*args)',
+            globals={'predict': prediction_method, 'args': (df.to_numpy(),)},
```
why do df.to_numpy()? I suspect some models will fail on this (e.g. if column order in train was different than what is supplied for test)

yromanyshyn replied:
hm/ do not why but I was thinking that it will fail if to provide a dataframe to the predict method, fixed

noamzbr reviewed (commented):
Also reminding to add example notebook, and to add to methodology suite with the condition

noamzbr commented on deepchecks/base/dataset.py:
```diff
@@ -297,6 +297,9 @@ def check_compatible_labels(self):
     # Validations:
 
+    # TODO: error messages is totally unrelated to the Dataclass itself
+    # I think we need to refactor this methods
```
You mean having the validations external to the Dataset? Possibly, they relate to the Dataset but not all datasets need to pass the same validations

yromanyshyn replied:
I mean that if I were the user of the library and if I called the `validate_index` method (outside of the check context) and got a message like this - `Check {check_name} requires dataset to have an index column`, I would have been confused. How check instances are related to the fact that I want to make sure that my dataset has an index?

the second thing that bothers me is that it does not help with type hints. Calling `validate_index` does not help the static type checker to infer that accessing `self.index_col()` will actually return `pd.Series` and not `Optional[pd.Series]`

noamzbr replied:
I get it, first point makes sense. @ItayGabbay WDYT? We can open a task for that next sprint

About type hints - is there a way to do in a way that helps? Note that type hints are more for us, they don't really help the user as far as I know

yromanyshyn replied:
> About type hints - is there a way to do in a way that helps? 

will try to create PR for that

> Note that type hints are more for us, they don't really help the user as far as I know

1. for users it means: better autocomplete, good documentation
2. I would say that they are not for us but for the static type checker 

benefits of extensively using type annotations (copied from mypy docs)
```
1.Static typing can make programs easier to understand and maintain. Type declarations can serve as machine-checked documentation. This is important as code is typically read much more often than modified, and this is especially important for large and complex programs.
2.Static typing can help you find bugs earlier and with less testing and debugging. Especially in large and complex projects this can be a major time-saver.
3.Static typing can help you find difficult-to-find bugs before your code goes into production. This can improve reliability and reduce the number of security issues.
```

noamzbr reviewed (changes requested):
Please add example notebook, and to add to methodology suite with the condition

yromanyshyn commented:
@noamzbr 

I added a notebook and added the check to the suite,
but there is a problem. Whenever you try to measure inference time, 
you always will get a little bit different result (different time)
and therefore the different output of the check. Because of this `make notebook` action fails.

Temporally I commented execution of the check within the notebook,
and the line that adds the check to the suite.

matanper commented on deepchecks/checks/methodology/model_inference_time.py:
```diff
@@ -0,0 +1,94 @@
+    """Measure model average inference time (in seconds) per sample.
+
+    Args:
+        number_of_samples (int): number of samples to use for inference
```
maximum number of samples (dataset might be smaller)

yromanyshyn replied:
updated

matanper commented on deepchecks/checks/methodology/model_inference_time.py:
```diff
@@ -0,0 +1,94 @@
+            DeepchecksValueError: If the 'test_dataset' is not a 'Dataset' instance with a label or
+                if 'model' is not a scikit-learn-compatible fitted estimator instance
+        """
+        return self._model_inference_time_check(dataset, model)
```
I'm not a fan of this split, is there a reason for it or just convention from other checks?

yromanyshyn replied:
ust convention from other checks

noamzbr replied:
not all checks do it I think. I think it should be reserved to cases when there is actual separation of function

matanper commented on deepchecks/checks/methodology/model_inference_time.py:
```diff
@@ -0,0 +1,94 @@
+        df = t.cast(pd.DataFrame, dataset.features_columns())
+
+        number_of_samples = len(df) if len(df) < self.number_of_samples else self.number_of_samples
+        df = df[:number_of_samples]
```
use dataframe sample method to take random samples (with random_state as seed)

yromanyshyn replied:
updated

matanper commented on deepchecks/checks/methodology/model_inference_time.py:
```diff
@@ -0,0 +1,94 @@
+            if avarage_time >= value:
+                return ConditionResult(False, details=(
+                    'Average model inference time for one sample (in seconds) '
+                    f'is greater than {format_number(value, floating_point=8)}'
```
```suggestion
                    f'is {format_number(avarage_time, floating_point=8)}'
```

matanper commented on tests/checks/methodology/model_inference_time_test.py:
```diff
@@ -0,0 +1,77 @@
+from tests.checks.utils import equal_condition_result, SCIENTIFIC_NOTATION_REGEXP
+
+
+def test_model_inference_time_check(
```
Add 
```
# Arrange
# Act
# Assert
```

yromanyshyn replied:
updated

matanper commented:
> @noamzbr
> 
> I added a notebook and added the check to the suite, but there is a problem. Whenever you try to measure inference time, you always will get a little bit different result (different time) and therefore the different output of the check. Because of this `make notebook` action fails.
> 
> Temporally I commented execution of the check within the notebook, and the line that adds the check to the suite.

Add to the problematic cells the next comment to skip the testing:
# NBVAL_SKIP

from https://nbval.readthedocs.io/en/latest/

matanper commented on deepchecks/checks/methodology/performance_overfit.py:
```diff
@@ -100,7 +100,7 @@ def plot_overfit():
                            display=[plot_overfit])
 
-    def add_condition_difference_not_greater_than(self, threshold: float) -> TD:
+    def add_condition_difference_not_greater_than(self: TD, threshold: float) -> TD:
```
isn't it weird to put type hint on self?

yromanyshyn replied:
nope, this is called Self-Type. This PEP gives a good explaining why we need to use this - [link](https://www.python.org/dev/peps/pep-0673/)
